Added dynamic fetching of networks from The Graph's network registry

GabrielCartier opened the pull request:
Updated the code to use the new types
Added a method for a custom network if needed.

GabrielCartier committed:
added pinax go lib to fetch networks

Updated the code to use the new types

GabrielCartier committed:
added add custom network method

GabrielCartier commented on codegen/chainconfig.go:
```diff
@@ -1,143 +1,47 @@
+		reg, err := registry.FromLatestVersion()
+		if err != nil {
+			// Fallback: try to load from local file
+			// TODO: Validate where to put the registry file
```
@maoueh just unsure where you'd want the file. Since the pinax lib has a method to fetch from a file, seems like the easiest solution to just use that?

maoueh replied:
A file cannot be packaged within the final built binary. So needs to be embedded within the binary itself using [go:embed](https://pkg.go.dev/embed) directive.

GabrielCartier commented on codegen/chainconfig.go:
```diff
@@ -1,143 +1,47 @@
+		}
+		m := make(map[string]*registry.Network)
+		for i, net := range reg.Networks {
+			if len(net.Services.Firehose) > 0 {
```
Do I also need to filter out the ones where `len(net.Services.Subtreams) > 0`?

maoueh replied:
It should be just `Substreams` actually, since we are in the Substreams service.

maoueh replied:
The `FirehoseEndpoint` used in the code was misleading.

maoueh commented on codegen/chainconfig.go:
```diff
@@ -1,143 +1,47 @@
+			reg, err = registry.FromFile("TheGraphNetworksRegistry.json")
+			if err != nil {
+				log.Fatalf("Failed to load registry from both network and file: %v", err)
+			}
```
We do not use the `log` package, just `panic` here

maoueh commented on codegen/chainconfig.go:
```diff
@@ -1,143 +1,47 @@
+// For now, this is a stub. To use, implement logic to add a *registry.Network to registryNetworks.
+func addCustomNetwork() {
+	// Example usage:
+	// registryNetworks["customnet"] = &registry.Network{ID: "customnet", ...}
```
The method should do the real thing. The best is to add `TRON` support right now for showcasing and testing purposes.

maoueh commented on codegen/project.go:
```diff
@@ -336,10 +336,12 @@ func (p *Project) GetRustOutputProtobufPath() string {
+	if net == nil {
 		return "", fmt.Errorf("network %q not found", p.Network)
 	}
-	return ChainConfigByID[p.Network].FirehoseEndpoint, nil
```
That is a bit unexpected, not impossible it was a mistake actually in first place, I'll need to double check.

maoueh replied:
Ok, reviewed quickly, I think `FirehoseEndpoint` was actually just a wrong variable name. Should be `Substreams[0]` here.

maoueh commented on codegen/chainconfig.go:
```diff
@@ -1,143 +1,47 @@
+)
+
+// GetRegistryNetworks returns a map of network ID to *registry.Network, using Pinax's registry as primary source.
+// Only networks with a Firehose endpoint are included.
```
```suggestion
// GetRegistryNetworks returns a map of network ID to *registry.Network, using The Graph's registry as primary source.
// Only networks with a Substreams endpoint are included.
```

maoueh commented on codegen/chainconfig.go:
```diff
@@ -1,143 +1,47 @@
+)
+
+var (
+	registryNetworks     map[string]*registry.Network
```
```suggestion
	registryNetworks     NetworkRegistry
```

maoueh commented on codegen/chainconfig.go:
```diff
@@ -1,143 +1,47 @@
+
+	registry "github.com/pinax-network/graph-networks-libs/packages/golang/lib"
+)
+
```
Let's use custom type for registry:

```
// NetworkRegistry is a thin wrapper around a [map[string]*registry.Network] to add some helper methods.
type NetworkRegistry map[string]*registry.Network
```

That way we will be able to attach methods to it.

maoueh commented on codegen/chainconfig.go:
```diff
@@ -1,143 +1,47 @@
-	},
+// addCustomNetwork can be used to add a custom network to the registry map for testing or development.
+// For now, this is a stub. To use, implement logic to add a *registry.Network to registryNetworks.
+func addCustomNetwork() {
```
Let's turn that into a method on `NetworkRegsitry`:

```suggestion
func (r NetworkRegistry) AddNetwork(network *registry.Network) {
```

maoueh commented on codegen/chainconfig.go:
```diff
@@ -1,143 +1,47 @@
+			}
+		}
+		registryNetworks = m
+		// Optionally add custom networks for testing
+		addCustomNetwork()
```
Let's add TRON right away

maoueh commented on codegen/chainconfig.go:
```diff
@@ -1,143 +1,47 @@
+// GetRegistryNetworks returns a map of network ID to *registry.Network, using Pinax's registry as primary source.
+// Only networks with a Firehose endpoint are included.
+// If fetching fails, it falls back to loading from a local JSON file ("TheGraphNetworksRegistry.json").
+func GetRegistryNetworks() map[string]*registry.Network {
```
```suggestion
func GetNetworkRegistry() NetworkRegistry {
```

maoueh commented on codegen/chainconfig.go:
```diff
@@ -1,143 +1,47 @@
 }
 
-var ChainConfigByID = map[string]*ChainConfig{
-	"mainnet": {
-		DisplayName:      "Ethereum Mainnet",
-		ExplorerLink:     "https://etherscan.io",
-		FirehoseEndpoint: "mainnet.eth.streamingfast.io:443",
```
Could you take all keys previously defined here and create a unit tests that ensures that all of them can be retrieved today so we ensure that code:

```
networks := GetRegistryNetworks()
	net := networks[p.Network]
```

Works for all previous cases.

GabrielCartier replied:
test fails becuause some of the `id` are different:
```
bnb -> bsc
polygon -> matic
amoy -> polygon-amoy
arbitrum -> arbitrum-one
```

But if we use the `Find` method it works. I guess the test is really to make sure all the stuff is there, so that's fine.

maoueh commented on codegen/project.go:
```diff
@@ -336,10 +336,12 @@ func (p *Project) GetRustOutputProtobufPath() string {
 
 func (p *Project) ChainEndpoint() (string, error) {
-	if ChainConfigByID[p.Network] == nil {
+	networks := GetRegistryNetworks()
+	net := networks[p.Network]
+	if net == nil {
```
We should define a `Find(key string)` on `NetworkRegistry` type, the method would first look for a network with the same ID, return it if found.

If no network with this ID is found, take all networks IDs, sort them, iterate in order the network IDs, for the network, inspect all aliases, if an alias matches, return it.

Add unit tests for this. This makes me think, tests could be flaky actually if the network registry is updated because by default we fetch from latest, let's see for now.

maoueh reviewed (changes requested):
Can you run some manual tests on the built CLI to ensure it still works correctly. The easiest is to go in `substreams-foundational-modules` and test `starknet` and `stellar`, both should by doing:

```
cd ethereum-common
substreams build
substreams run ethereum-common-v0.3.2.spkg -t +10
# Print 10 empty blocks, complains about no network field

cd ../mantra-common
substreams build
substreams run mantra-common-v0.1.1.spkg -t +10

cd ../stellar-common
substreams build
substreams run stellar-foundational-v0.3.0.spkg -s 55411000 -t +10
```

Also right now `stellar-common` doesn't work, but it should. The problem today with Stellar is that in the foundational modules, the network is defined as `stellar-mainnet` but the ID is actually `stellar` so it's not found.

The `Find` method I talked about in the PR comment should solve this as it would check also aliases where `stellar-mainnet` is defined.

GabrielCartier committed:
PR nits

moved chainconfig to it's own package to reuse
added tests
embedded json file
added helper methods and types

GabrielCartier committed:
added method to fetch specific networks

GabrielCartier commented on networks/chainconfig.go:
```diff
@@ -0,0 +1,141 @@
+		registryNetworks = m
+
+		// Add custom Tron network until it's added to the registry
+		registryNetworks.addCustomNetwork(&registry.Network{
```
Not sure if all the data is valid here, I believe so?

maoueh replied:
Seems quite good, I will probably task you to open a PR for this on the network registry

GabrielCartier replied:
For the PR for network registry, do I need to add more stuff? How do I gather the missing stuff?

maoueh replied:
List me the missing pieces and I'll ask TRON team

GabrielCartier committed:
removed HardcodedEndpoints

GabrielCartier committed:
added find by genesis and modified find

GabrielCartier commented:
> Can you run some manual tests on the built CLI to ensure it still works correctly. The easiest is to go in `substreams-foundational-modules` and test `starknet` and `stellar`, both should by doing:
> 
> ```
> cd ethereum-common
> substreams build
> substreams run ethereum-common-v0.3.2.spkg -t +10
> # Print 10 empty blocks, complains about no network field
> 
> cd ../mantra-common
> substreams build
> substreams run mantra-common-v0.1.1.spkg -t +10
> 
> cd ../stellar-common
> substreams build
> substreams run stellar-foundational-v0.3.0.spkg -s 55411000 -t +10
> ```
> 
> Also right now `stellar-common` doesn't work, but it should. The problem today with Stellar is that in the foundational modules, the network is defined as `stellar-mainnet` but the ID is actually `stellar` so it's not found.
> 
> The `Find` method I talked about in the PR comment should solve this as it would check also aliases where `stellar-mainnet` is defined.

Tested it, all seems working. Stellar is also working properly.

GabrielCartier commented on codegen/project.go:
```diff
@@ -336,10 +337,12 @@ func (p *Project) GetRustOutputProtobufPath() string {
 func (p *Project) ChainEndpoint() (string, error) {
-	if ChainConfigByID[p.Network] == nil {
+	networks := networks.GetRegistryNetworksWithSubstreams()
+	net := networks[p.Network]
```
Do we want to use `Find` here or do we assume id is correct?

maoueh replied:
Yeah use Find

maoueh commented on networks/TheGraphNetworksRegistry.json:
```diff
@@ -0,0 +1,5538 @@
+{
```
Let's rename the actual file so it as the version in it and fallback, so the file should be named `fallback_TheGraphNetworkRegistry_0.6.0.json`.

Can you also add a quick bash script in `networks` named `update-fallback-registry.sh` that:
- Record actual filename in a variable (for deletion)
- Fetches the latest version using `curl` and extract version using `jq` and write that to correct new name
- Delete the old filename if everything went well.

GabrielCartier replied:
Also added updating the `//go:emded` directive to point to the proper file.

maoueh commented on networks/chainconfig.go:
```diff
@@ -0,0 +1,152 @@
+	if n, ok := r[key]; ok {
+		return n
+	}
+	ids := make([]string, 0, len(r))
+	for id := range r {
+		ids = append(ids, id)
+	}
```
```suggestion
ids := slices.Collect(maps.Keys(r))
```

maoueh commented on networks/chainconfig.go:
```diff
@@ -0,0 +1,152 @@
+}
+
+// GetRegistryNetworksWithSubstreams returns only networks with Substreams endpoints.
+func GetRegistryNetworksWithSubstreams() NetworkRegistry {
```
```suggestion
func GetSubstreamsRegistry() NetworkRegistry {
```

Will be good enough with the documentation you have.

maoueh commented on networks/chainconfig.go:
```diff
@@ -0,0 +1,152 @@
+	return filtered
+}
+
+// GetRegistryNetworksWithFirehose returns only networks with Firehose endpoints.
+func GetRegistryNetworksWithFirehose() NetworkRegistry {
```
Same `GetFirehoseRegistry`

maoueh commented on tui2/pages/output/output.go:
```diff
@@ -75,7 +76,7 @@ func New(c common.Common, config *request.Config) (*Output, error) {
 	getBytesEncodingPerNetwork := func(endpoint string) dynamic.BytesRepresentation {
 		bytesAsBase58Chains := []string{"solana-mainnet-beta", "solana-mainnet", "solana-devnet"}
 		for _, chain := range bytesAsBase58Chains {
-			if manifest.HardcodedEndpoints[chain] == endpoint {
+			if networks.GetRegistryNetworksWithSubstreams().Find(chain).Services.Substreams[0] == endpoint {
 				return dynamic.BytesAsBase58
 			}
 		}
```
The whole `getBytesEncodingPerNetwork` should be refactored actually while reviewing this. There is no good reason to check just for Solana here. The registry can determined network desired bytes representation via `Firehose.BytesEncoding`. 

Can you move that function to `networks` package `GetBytesRepresentation(key string)`. The `config.Endpoint` lookup is a bit weird also, because `config.Pkg.Network` is better to try first, but it's fair to then try by endpoint if `network`  is not defined and unfindable. Rules:

- Use existing network search rules
- If nothing found, make an endpoint search like now

maoueh replied:
This code would then be refactored something like:

```
lookupKey := config.Pkg.Network
if lookupKey == "" { lookupKey = config.Endpoint }

bytesRepresentation := networks.GetBytesRepresentation(lookupKey)
```

GabrielCartier replied:
Ok, I didn't implement exactly how you said to try to keep network as "clean" as possible.
I added a `FindBySubstreamsEndpoint` method and `GetBytesEncoding` in `chainconfig.go`.
Then in `output.go` I try to find the network using `config.Pkg.Network` if it's set, else I try to find by endpoint. I then convert the result in `dynamic.BytesRepresentation` which is the type used in `output.go`.
It defaults to `hex` as before.

GabrielCartier replied:
Note, I added the `bytesEncodingToReprensatation` to `tui2/common/utils` because it's also used in `tui.go`. Not sure we want it there, I'll wait for your advice on this.

GabrielCartier committed:
Merge branch 'develop' into feature/pinax-networks

GabrielCartier committed:
PR Nits

renamed functions
Added script to update fallback registry
Added a method to find network by endpoint
Added a method to get bytes encoding from a network
cleaned output.go to work with new code

GabrielCartier committed:
fixed tests

maoueh commented on networks/update-fallback-registry.sh:
```diff
@@ -0,0 +1,53 @@
+  echo "Deleted old fallback file: $old_file"
+fi
+
+# Update go:embed line in chainconfig.go
+sed -i '' "s|^//go:embed fallback_TheGraphNetworkRegistry_.*\\.json$|//go:embed $new_file|" chainconfig.go
```
Overachiever 👍
